File: citation.py
#!/bin/env python3
from PIL import Image, ImageDraw, ImageFont
import math
import argparse
import re
import colors

m = 2
w, h = 183, 80
stamp_name = "moa.png"
main_font = "BMmini.ttf"
use_alt_font = False



parser = argparse.ArgumentParser(description='Generates Papers Please citation cards.')
parser.add_argument('lines', nargs=1,
                   help='lines of the citation, separated by ;')
parser.add_argument('penalty', nargs="?", default="No penalty - warning issued",
                   help='penalty text')
parser.add_argument('--title', '-t', default="M.O.A. Citation",
                   help='title of the citation (default "M.O.A. CITATION")')
parser.add_argument('--stamp', '-s', default="moa.png",
                   help='name of stamp .png file (default: moa)')
parser.add_argument('--altfont', action="store_true",
                   help='use alternative (old) font for the title')
parser.add_argument('--condensed', action="store_true",
                   help='use smaller spacing for lines')
parser.add_argument('--barcode', default="1,1,1,2,2",
                   help='barcode format to use in widths of lines as comma separated list')
parser.add_argument('--theme', '-c', default = "default",
                   help='color scheme - can be overriden by individual color settings',
                   choices = [x for x in dir(colors) if x[0:2] != '__'])
parser.add_argument('--foreground', '--fg', default = None,
                   help='hex color to use for foreground text and elements')
parser.add_argument('--background', '--bg', default = None,
                   help='hex color to use for background')
parser.add_argument('--details', '--det', default = None,
                   help='hex color to use for background details')

args = parser.parse_args()
stamp_name = args.stamp
lines = args.lines[0].split(";")
use_alt_font = args.altfont
condensed = args.condensed
theme = getattr(colors, args.theme)
c_bg   = args.background or theme[0]
c_fg   = args.foreground or theme[1]
c_elem = args.details    or theme[2]

minlines = 3
lineheight = 10
if condensed:
    minlines = 4
    lineheight = 9

def line_bind(text):
    max_w = (w - 11 - 12) * m
    size = bm.getsize(text)
    if size[0] <= max_w:
        return (text, "")
    chars = 1
    size = bm.getsize(text[:chars])
    while size[0] < max_w:
        size = bm.getsize(text[:chars])
        chars += 1
    return (text[:chars - 1], text[chars - 1:])

bm = ImageFont.truetype(main_font, 16)

newlines = []
for x in lines:
    newlines.extend(x.split("\n"))
lines = newlines
newlines = []
for x in lines:
    out = line_bind(x)
    newlines.append(out[0])
    while len(out[1]) != 0:
        out = line_bind(out[1])
        newlines.append(out[0])
lines = newlines 

if len(lines) > minlines:
    h += (len(lines) - minlines) * lineheight


im = Image.new("RGB",(w*m, h*m), c_bg)
draw = ImageDraw.Draw(im)
if use_alt_font == True:
    global tbm
    tbm = ImageFont.truetype("Megan_Serif.ttf", 16)

def dots(line, offset, color = c_elem):
    start = offset[0]
    end   = w - 1 - offset[1]
    num = 0
    for x in range((end - start + 1) // 2 + 1):
        c = [start*m + num*m*2, line*m, start*m + num*m*2 + m - 1, line*m + m - 1]
        draw.rectangle(c, color)
        num += 1
    return


def rect(x, y, w, h, color = c_elem):
    draw.rectangle([x * m, y * m, (x + w) * m - 1, (y + h) * m - 1], color)


def roll():
    for x in range((h + 3) // 9):
        rect(2, 3 + 9 * x, 3, 3)
        rect(w - 7, 3 + 9 * x, 3, 3)


def line(on, text, color=c_fg):
    offset = 4
    remaining = ""
    if on != 0:
        offset = 22 + (on - 1) * lineheight

    if on == 0 and use_alt_font == True:
        offset = 3
        draw.text((11*m - 1, offset*m), text.replace(" ", "    "), font=tbm, fill=color)
    else:
        draw.text((11*m - 1, offset*m), text, font=bm, fill=color)
    return remaining

def stamp(): # 32x32 at 150x44
    stamp_bg = Image.open("stamp_bg.png")
    draw.bitmap((150, (h - 4 - 32) *m), stamp_bg, c_elem)
    stamp_bg.close()
    stamp = Image.open(stamp_name)
    draw.bitmap((150, (h - 4 - 32) *m), stamp, c_bg)
    stamp.close()


def footer(text, color=c_fg):
    text = text.upper()
    y = (h - 15) * m
    size = bm.getsize(text)

    x = math.ceil((w - 6 - 10 - (size[0] - 1) // m) / 2) * m + 6 * m
    draw.text((x - 1, y), text, font=bm, fill=color)
    return

def barcode(spec, color=c_fg):
    end = 172
    width = 2
    for x in spec:
        width += 1 + x
    offset = 0
    for x in reversed(spec):
        rect(end - offset - x, 3, x, 6, color)
        offset += 1 + x
    rect(end - offset - 2, 3, 2, 3, color)



stamp()
dots(0, (0, 0))
roll()
rect(w - 1, 0, w, h, c_elem)
dots(h - 1, (1, 0))

barcode([int(i) for i in re.findall(r"[\w']+", args.barcode)])

dots(17, (8, 10), c_fg)
line(0, args.title.upper())

offset = 1
for x in range(len(lines)):
    left = line(x + offset, lines[x])
    while len(left) != 0:
        offset += 1
        left = line(x + offset, left)

if condensed:
    dots(h - 22 - 1, (8, 10), c_fg)
else:
    dots(h - 26 - 1, (8, 10), c_fg)
footer(args.penalty)

im.save("citation.png", "PNG")

File: web/citation.ts
let citation = null;
class Colors {
    constructor(public background, public foreground, public secondary) {

    };
}

class Bitmap {
    constructor(public data, public width, public height) {}

    draw(cit: Citation, x, y): void {
        for (let n = 0; n < this.width * this.height; n++) {
            if (this.data[n]) {
                cit.rect(x + n % this.width, y + Math.floor(n / this.width), 1, 1);
            }
        }
    };
}

class Citation {
    mainfont:string;
    altfont: string;
    multiplier:number = 2;
    width:number = 183;
    height:number = 80;

    condensed = true;
    use_alt_font = false;
    title = "";
    penalty = "";
    lines: string[] = [];

    minlines;
    lineheight;

    canvas: HTMLCanvasElement;
    context: CanvasRenderingContext2D;
    stamp: Bitmap = new Bitmap(
        [
            0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,
            0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,
            0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,
            0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,
            0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,
            0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,
            0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,
            0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,
            0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,
            0,1,1,1,1,1,1,1,1,0,1,1,1,1,1,0,0,1,1,1,1,1,0,1,1,1,1,1,1,1,1,0,
            0,1,1,1,1,1,1,1,1,0,0,1,1,1,0,0,0,0,1,1,1,0,0,1,1,1,1,1,1,1,1,0,
            1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,
            1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,
            1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,1,1,0,0,0,1,1,1,1,1,1,1,1,1,
            1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,1,1,1,0,0,0,1,1,1,1,1,1,1,1,1,
            1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,1,1,1,0,0,0,0,1,1,1,1,1,1,1,1,1,
            1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,1,1,1,0,0,0,0,0,1,1,1,1,1,1,1,1,1,
            1,1,1,1,1,1,1,1,1,0,0,0,0,0,1,1,1,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,
            1,1,1,1,1,1,1,1,1,0,0,0,0,1,1,1,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,
            1,1,1,1,1,1,1,1,1,1,0,0,1,1,1,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,
            0,1,1,1,1,1,1,1,1,1,0,0,1,1,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,0,
            0,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,0,
            0,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,0,
            0,0,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,0,0,
            0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,
            0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,
            0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,
            0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,
            0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,
            0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,
            0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,
            0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0
        ], 32, 32);

    color: Colors;
    private barcode_data: number[] = [1,1,1,2,2];

    constructor() {
        this.canvas = <HTMLCanvasElement>document.getElementById("main_canvas");
        this.context = this.canvas.getContext('2d');
        this.context.imageSmoothingEnabled = false;
        let form = document.getElementById("form");
        form.addEventListener("submit", (event)=>{event.preventDefault();this.refresh(); this.render(); return false});
        form.addEventListener("change", (event)=>{event.preventDefault();this.refresh(); this.render(); return false});

        this.color = new Colors("#f3d7e6", "#5a5559", "#bfa8a8");
        this.refresh();
        this.render();
    }

    refresh() {
        let fontsize = "" + (8 * this.multiplier) + "px ";
        this.mainfont = fontsize + "BMmini";
        this.altfont = fontsize + "\"Megan Serif\"";
        this.context.font = this.mainfont;

        let text = (<HTMLTextAreaElement>document.getElementsByName("text")[0]).value;
        this.penalty = (<HTMLInputElement>document.getElementsByName("penalty")[0]).value;
        this.title = (<HTMLInputElement>document.getElementsByName("title")[0]).value;
        this.condensed = (<HTMLInputElement>document.getElementsByName("condensed")[0]).checked;
        this.color.foreground = (<HTMLInputElement>document.getElementsByName("foreground")[0]).value;
        this.color.background = (<HTMLInputElement>document.getElementsByName("background")[0]).value;
        this.color.secondary = (<HTMLInputElement>document.getElementsByName("secondary")[0]).value;
        this.use_alt_font = (<HTMLInputElement>document.getElementsByName("altfont")[0]).checked;
        this.barcode_data = (<HTMLInputElement>document.getElementsByName("barcode")[0]).value
            .split(/[ ,;./|=+]/)
            .filter((a) => {return parseInt(a)})
            .map((a) => {return parseInt(a)});


        this.minlines = this.condensed ? 4 : 3;
        this.lineheight = this.condensed ? 9 : 10;


        let tmpLines = text.split("\n");
        this.lines = [];
        for (let x of tmpLines) {
            let out = this.wrapLine(x);
            this.lines.push(out[0]);
            while (out[1].length != 0) {
                out = this.wrapLine(out[1]);
                this.lines.push(out[0]);
            }
        }

        if (this.lines.length > this.minlines) {
            this.height = (80 + (this.lines.length - this.minlines) * this.lineheight);
            console.log(this.height, this.lines.length);
        } else {
            this.height = 80;
        }

        this.canvas.height = this.height * this.multiplier;
        this.canvas.width = this.width * this.multiplier;
    }

    rect(x, y, w, h) {
        return this.context.fillRect(x * this.multiplier, y * this.multiplier, w * this.multiplier, h * this.multiplier);
    }

    line(on, text) {
        this.context.fillStyle = this.color.foreground;
        let offset = 4;
        if (on != 0) {
            offset = 22 + (on - 1) * this.lineheight;
        }

        if (on == 0 && this.use_alt_font) {
            offset = 3 + 8;
            this.context.font = this.altfont;
            this.context.fillText(text.replace(" ", "    "), 11 * this.multiplier, offset * this.multiplier);
            this.context.font = this.mainfont;
        } else {
            this.context.font = this.mainfont;
            this.context.textBaseline = 'hanging';
            this.context.fillText(text, 11 * this.multiplier, offset * this.multiplier);
        }
    };

    footer(text: string, color=this.color.foreground) {
        text = text.toUpperCase();
        let y = (this.height - 15) * this.multiplier;
        let size = this.context.measureText(text);

        let x = (Math.ceil((this.width - 6 - 10 - Math.floor((size.width - 1) / this.multiplier)) / 2) + 6)* this.multiplier;
        this.context.fillText(text, x, y);
        return
    }

    dots(line: number, offset: number[], color:string=this.color.secondary) {
        this.context.fillStyle = color;
        let start = offset[0];
        let end   = this.width - 1 - offset[1];
        for (let x = 0; x < Math.floor((end - start + 1) / 2) + 1; x++) {
            this.rect(start + x*2, line, 1, 1);
        }
        return
    }

    barcode(data) {
        let end = 172;
        let width = 2;
        for (let x of data) {
            width += 1 + x
        }
        let offset = 0;
        this.context.fillStyle = this.color.foreground;
        for (let i = 0; i < data.length; i++) {
            let x = data[data.length - 1 - i];
            this.rect(end - offset - x, 3, x, 6);
            offset += 1 + x;
        }
        this.rect(end - offset - 2, 3, 2, 3);
    }

    render() {
        this.context.fillStyle = this.color.background;
        this.rect(0, 0, this.width, this.height);
        this.context.fillStyle = this.color.secondary;
        this.stamp.draw(this, 75, this.height - 4 - 32);
        this.dots(0, [0, 0]);
        this.roll();
        this.context.fillStyle = this.color.secondary;
        this.rect(this.width - 1, 0, this.width, this.height);
        this.dots(this.height - 1, [1, 0]);

        this.barcode(this.barcode_data);

        this.dots(17, [8, 10], this.color.foreground);
        this.line(0, this.title.toUpperCase());

        for (let x=0; x < this.lines.length; x++) {
            this.line(x + 1, this.lines[x]);
        }

        if (this.condensed) {
            this.dots(this.height - 22 - 1, [8, 10], this.color.foreground)
        } else {
            this.dots(this.height - 26 - 1, [8, 10], this.color.foreground)
        }
        this.footer(this.penalty)
    }

    roll() {
        for (let x=0; x < Math.floor((this.height + 3) / 9); x++) {
            this.rect(2, 3 + 9 * x, 3, 3);
            this.rect(this.width - 7, 3 + 9 * x, 3, 3);
        }
    }

    wrapLine(text: string): string[] {
        let max_w = (this.width - 11 - 12) * this.multiplier;
        let size = this.context.measureText(text);
        if (size.width <= max_w) {
            return [text, ""];
        }
        let chars = 1;
        size = this.context.measureText(text.substring(0, chars));
        while (size.width < max_w) {
            size = this.context.measureText(text.substring(0, chars));
            chars += 1;
        }
        return [text.substring(0, chars - 1), text.substring(chars - 1)]
    }

    static init() {
        citation = new Citation();
    }
}


document.addEventListener("DOMContentLoaded", Citation.init, false);
